phylo/src/optimisers/topo_optimiser.rs

```diff
@@ -0,0 +1,122 @@
+                    continue;
+                }
+                let regraft_locations = Self::find_regraft_options(prune_branch, &info);
+                let mut moves: Vec<(f64, Tree)> = Vec::with_capacity(regraft_locations.len());
+                for regraft_branch in &regraft_locations {
+                    let mut new_info = info.clone();
+                    new_info.tree = info.tree.rooted_spr(prune_branch, regraft_branch).unwrap();
+                    let mut logl = self.model.cost(&new_info, false);
+                    if logl <= curr_cost {
+                        // reoptimise branch length at the regraft location
+                        let o = BranchOptimiser::new(self.model, &new_info);
+                        let (blen_logl, blen) = o.optimise_branch(regraft_branch, &new_info)?;
+                        if blen_logl > logl {
+                            new_info.tree.set_blen(regraft_branch, blen);
+                            logl = blen_logl;
+                        }
+                    }
+                    moves.push((logl, new_info.tree.clone()));
+                }
+                let (best_logl, best_tree) = moves
+                    .into_iter()
+                    .max_by(|(a, _), (b, _)| a.partial_cmp(b).unwrap())
+                    .unwrap();
+                if best_logl > curr_cost {
+                    curr_cost = best_logl;
+                    info.tree = best_tree;
+                    debug!("Regrafted {} with new logl {}.", prune_branch, curr_cost);
+                } else {
+                    debug!(
+                        "No improvement regrafting {}, best logl {}.",
+                        prune_branch, best_logl
+                    );
+                }
+                // Optimise branch lengths on current tree to match PhyML
+                let o = BranchOptimiser::new(self.model, &info).run()?;
+                if o.final_logl > curr_cost {
+                    curr_cost = o.final_logl;
+                    info = o.i;
+                }
+                info.tree.clean(true);
+            }
+        }
+
+        info!(
+            "Final logl: {}, achieved in {} iteration(s).",
+            curr_cost, iterations
+        );
+        Ok(PhyloOptimisationResult {
+            initial_logl,
+            final_logl: curr_cost,
+            iterations,
+            i: info,
+        })
+    }
+}
+
+impl<EM: PhyloCostFunction> TopologyOptimiser<'_, EM> {
+    fn find_regraft_options(prune_branch: &NodeIdx, info: &PhyloInfo) -> Vec<NodeIdx> {
+        let all_locations = info.tree.preorder();
+        let prune_subtrees = info.tree.preorder_subroot(prune_branch);
+        let sibling = &info.tree.sibling(prune_branch).unwrap();
+        let parent = &info.tree.node(prune_branch).parent.unwrap();
+        all_locations
+            .iter()
+            .filter(|&n| {
+                !prune_subtrees.contains(n) && n != sibling && n != parent && n != &info.tree.root
```

Comment:
I recommend changing the order of the checks to ensure fast checks are done first

Reply:
Done